### PR TITLE
add support robot-framework

### PR DESCRIPTION
**Pull request by dalanicolai**
Here is a rudimentary implementation of the robot-framework lsp-client. It can already be used in a quite straightforward way, so you might agree to merge it already. However, it would be nice to at least implement the right menu's for the log and trace variables. So for now I am sending this PR mainly so you can have a look at it. I can update the documentation too, but I don't know how to instal the server without vscode, I never really worked with npm or javascript. Would be great to get your feedback.

Additionally I could update the spacemacs lsp layer later

### CODE REVIEW

File: lsp-clients.el [hunk 1]
```diff
@@ -828,5 +830,120 @@ responsiveness at the cost of possible stability issues."
...
+  :type 'boolean
+  :group 'lsp-rf)
+
+(defcustom lsp-rf-language-server-dir "~/.vscode/extensions/tomiturtiainen.rf-intellisense-2.8.0/server/library-docs/"
+  "Libraries directory for libraries in lsp-rf-language-server-libraries"
+  :type 'string
+  :group 'lsp-rf)
+
+(defcustom lsp-rf-language-server-libraries '("BuiltIn-3.1.1" "Collections-3.0.4")
+  "Libraries whose keywords are suggested with auto-complete"
```
Comment (yyoncho) [inline lsp-clients.el]: should be vector. Note that emacs27+ json serialization treats list as objects

File: lsp-clients.el [hunk 2]
```diff
@@ -828,5 +830,120 @@ responsiveness at the cost of possible stability issues."
...
+
+(lsp-register-custom-settings
+ '(
+   ("rfLanguageServer.trace.server" lsp-rf-language-server-trace-server)
+   ("rfLanguageServer.logLevel" lsp-rf-language-server-log-level)
+   ("rfLanguageServer.libraries" lsp-rf-language-server-libraries)
+   ("rfLanguageServer.pythonKeywords" lsp-rf-language-server-python-keywords t)
+   ("rfLanguageServer.excludePaths" lsp-rf-language-server-exclude-paths)
+   ("rfLanguageServer.includePaths" lsp-rf-language-server-include-paths)
+   ))
```
Comment (yyoncho) [inline lsp-clients.el]: close on  the same line

File: lsp-clients.el [hunk 3]
```diff
@@ -828,5 +830,120 @@ responsiveness at the cost of possible stability issues."
...
+                  :initialized-fn (lambda (workspace)
+                                    (with-lsp-workspace workspace
+                                      (lsp--set-configuration
+                                       (lsp-configuration-section "rfLanguageServer"))
+                                      (lsp-request "buildFromFiles"
+                                                   (list :files
+                                                         (vconcat
+                                                          (rf-language-server-exclude-paths
+                                                           (directory-files-recursively
+                                                            (projectile-project-root)
```
Comment (yyoncho) [inline lsp-clients.el]: You should not use projectile here but `lsp--workspace-root`

File: lsp-clients.el [hunk 4]
```diff
@@ -828,5 +830,120 @@ responsiveness at the cost of possible stability issues."
...
+   ("rfLanguageServer.logLevel" lsp-rf-language-server-log-level)
+   ("rfLanguageServer.libraries" lsp-rf-language-server-libraries)
+   ("rfLanguageServer.pythonKeywords" lsp-rf-language-server-python-keywords t)
+   ("rfLanguageServer.excludePaths" lsp-rf-language-server-exclude-paths)
+   ("rfLanguageServer.includePaths" lsp-rf-language-server-include-paths)
+   ))
+
+(lsp-register-client
+ (make-lsp-client :new-connection (lsp-stdio-connection
+                                   `("node" ,(expand-file-name
```
Comment (yyoncho) [inline lsp-clients.el]: please make `node` part of the command to allow customizing it.

File: lsp-clients.el [hunk 5]
```diff
@@ -828,5 +830,120 @@ responsiveness at the cost of possible stability issues."
...
+  "An array of files that should be included by the parser. Glob patterns as strings are accepted (eg. *.robot between double quotes)"
+  :type 'lsp-string-vector
+  :group 'lsp-rf)
+
+(defcustom lsp-rf-language-server-exclude-paths []
+  "An array of files that should be ignored by the parser. Glob patterns as strings are accepted (eg. *bad.robot between double quotes)"
+  :type 'lsp-string-vector
+  :group 'lsp-rf)
+
+(defcustom lsp-rf-language-server-python-keywords nil
```
Comment (yyoncho) [inline lsp-clients.el]: if it is deprecated consider not adding it at all.

File: lsp-clients.el (hunk 1 above)
Comment (dalanicolai) [reply lsp-clients.el]: This means that for now it works correctly, but from emacs 27+ on it will fail?

Comment (yyoncho) [reply lsp-clients.el]: yes.

File: lsp-clients.el (hunk 3 above)
Comment (dalanicolai) [reply lsp-clients.el]: I've added lsp-workspace-root. I looked at lsp--workspace-root but I am not familiar with cl structures yet. I don't know what to use as argument (CL-X). I tried lsp--cur-workspace but its value is nil

File: lsp-clients.el [hunk 6]
```diff
@@ -828,5 +830,120 @@ responsiveness at the cost of possible stability issues."
...
+                  :initialized-fn (lambda (workspace)
+                                    (with-lsp-workspace workspace
+                                      (lsp--set-configuration
+                                       (lsp-configuration-section "rfLanguageServer"))
+                                      (lsp-request "buildFromFiles"
+                                                   (list :files
+                                                         (vconcat
+                                                          (parse-rf-language-server-exclude-paths
+                                                           (directory-files-recursively
+                                                            (lsp-workspace-root)
```
Comment (yyoncho) [inline lsp-clients.el]: pass `workspace`

Comment (dalanicolai) [reply lsp-clients.el]: What do you mean? You mean pass that to the `lsp--workspace-root`? (Probably not because that doesn't work)

Comment (yyoncho) [reply lsp-clients.el]: should be `(lsp--workspace-root workspace)`

Comment (dalanicolai) [reply lsp-clients.el]: Okay, replaced that line. However, I tested this before from a buffer that should be "inside" the workspace but then workspace was void. Now I see the workspace is passed by the initialized-fn "keyword".

**Comment by yyoncho** [comment]
Thank you for contributing to `lsp-mode` and congrats for your first PR! :tada:

**Comment by dalanicolai** [comment]
Great! Thanks... I will soon have a look where to update some documentation and probably also add some configuration for the spacemacs layer...

**Pull request merged**